### PR TITLE
Javalab order persists

### PR DESCRIPTION
**Pull request by fisher-alice**
This work follows up [this PR](https://github.com/code-dot-org/code-dot-org/pull/48139) which added the options to move file tabs to the left or right in the file dropdown menu.

When the browser was reloaded or when the user moved to another level, the order of the file tabs did not persist.
This PR allows changes in the file tab order to persist. In order to accomplish this, I added a key called 'order' to the files stored in the state value 'sources'. The key 'order' is an integer which designates the order the file tabs appear in the Java Lab editor. The leftmost file tab has order 0, the next file tab has order 1, ... Whenever the user rearranges the order of the tabs, the orders are updated in sources correspondingly. 

## Links
[jira - When file tabs are rearranged in the editor, order of tabs persist](https://codedotorg.atlassian.net/browse/JAVA-686)
[spec](https://docs.google.com/document/d/1TNIZwpGWLpB3P5IY7nJ8uHbvgNbzsxaeRSqtGzuSmEw/edit#heading=h.xtximbbqb4op)

Screencast Videos
Before update:

https://user-images.githubusercontent.com/107423305/191085831-e4f65bc6-0e46-48c4-8544-c1b023486ecd.mp4

After update: Order of file tabs persists after browser reload - added and deleted files

https://user-images.githubusercontent.com/107423305/191087532-e84803e3-1087-46ad-84d3-3754c0e63077.mp4

After update: Order of file tabs persists after user leave level and returns

https://user-images.githubusercontent.com/107423305/191088956-2dcfc891-4c64-4a6b-8c61-eb23e232ed9a.mp4

After update: Added files from Backpack, deleted files on a validation level

https://user-images.githubusercontent.com/107423305/191285230-1620a6ec-f73f-4071-bbfd-9918b2121200.mp4

After update: In levelbuilder in start mode - added validation and support files and another starter file and reordered file tabs - order persisted. Then showed corresponding level from student view - only starter files appeared in same order as in levelbuilder start mode but excluding the validation and support files.
Reordering student tabs does not impact levelbuilder start mode ordering of tabs.

https://user-images.githubusercontent.com/107423305/191570879-53e36671-5beb-406b-a756-44717cd81a13.mp4

## Testing story
- Updated unit tests in JavalabEditorTest.js by adding the key 'order' to the files stored in sources.
- Added 2 new unit tests to check if sources are updated correctly when files are moved and when a file is deleted.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by fisher-alice — COMMENTED** [review]
Made a copy of sources instead of modifying sources in javalabRedux.js line 498 in response to @molly-moen 's [feedback](https://github.com/code-dot-org/code-dot-org/pull/48171#discussion_r974683832)

Added a comment clarifying role of orderOfFiles in JavalabFileHelper.js beginning on line 24 in response to @molly-moen 's [feedback](https://github.com/code-dot-org/code-dot-org/pull/48171#discussion_r974686315)
Thanks Molly!

FYI - created a new version of this branch for part 2 of this project because [old version](https://github.com/code-dot-org/code-dot-org/pull/48171) was not showing diffs between part 1 and part 2 correctly.

File: apps/src/javalab/JavalabFileHelper.js [hunk 1]
```diff
@@ -20,17 +20,53 @@ export const getTabKey = index => `file-${index}`;
...
   let orderedTabKeys = [];
-
+  let unorderedTabKeys = [];
+  // sources contains file information as key-value pairs.
+  // The key is the file name such as 'MyClass.java'.
+  // The value is a file object with the keys: text (source code),
+  // order (order of the file tab in orderedTabs from 0 to number of files - 1),
+  // isVisible, and isValidation.
+  // The orderOfFiles array will contain the 'order' of each file as they are assigned
+  // their tabKey names (such as 'file-0', 'file-1') which is stored in fileMetadata
```
Comment (molly-moen) [inline apps/src/javalab/JavalabFileHelper.js]: Clarification: does `orderOfFiles` look something like `[1, 0, 3, 2]`, where `1` == index 1 in `unorderedTabKeys`? Since this setup is somewhat complicated it would be good to have an example of the array in the comment so it's clear for future readers :)

**Review by molly-moen — APPROVED** [review]
Looks good, just a comment on a comment :)

File: apps/src/javalab/javalabRedux.js [hunk 2]
```diff
@@ -58,7 +62,7 @@ const INCREASE_EDITOR_FONT_SIZE = 'javalab/INCREASE_EDITOR_FONT_SIZE';
 const DECREASE_EDITOR_FONT_SIZE = 'javalab/DECREASE_EDITOR_FONT_SIZE';
 
 const initialSources = {
-  'MyClass.java': {text: '', isVisible: true, isValidation: false}
+  'MyClass.java': {text: '', order: 0, isVisible: true, isValidation: false}
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/javalabRedux.js]: Do you think `tabIndex` or `tabOrder` would be clearer? I think `order` is probably fine too, but it may not be immediately obvious that this refers to the order of the file tabs in the editor, especially when viewing this data outside of the context of this code (ex. viewing JSON directly in S3).

Comment (fisher-alice) [reply apps/src/javalab/javalabRedux.js]: Yes - I agree that would be an improvement. Will change order to tabOrder - thanks!

File: apps/src/javalab/JavalabFileHelper.js (hunk 1 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: Yes, that's correct - and I have added detailed comments including an example. Thanks so much Molly!

File: apps/src/javalab/JavalabFileHelper.js [hunk 3]
```diff
@@ -18,20 +18,80 @@ export function getDefaultFileContents(filename, viewMode) {
...
+  The value is a file object with the keys: text (source code),
+  tabOrder (order of the file tab in orderedTabs from 0 to (number of files - 1)),
+  isVisible, and isValidation (default is false)
+  The orderOfFiles array will contain the 'fileOrder' of each file as they are assigned
+  their tabKey names (such as 'file-0', 'file-1') which is stored in fileMetadata.
+  For example, if sources contains the following:
+  {
+    'Class1.java': {text: '', tabOrder: 2, isVisible: true},
+    'Class2.java': {text: '', tabOrder: 0, isVisible: true},
+    'Class3.java': {text: '', tabOrder: 1, isVisible: true}
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabFileHelper.js]: Did you consider having the file order in the project json file change rather than adding a new key to the file? It seems like that approach might be more straightforward rather than having to assign metadata and orderedTabKeys after the fact.

File: apps/src/javalab/JavalabFileHelper.js [hunk 4]
```diff
@@ -18,20 +18,80 @@ export function getDefaultFileContents(filename, viewMode) {
...
+  },
+  then depending on how Object.keys iterates through the keys on sources, we could have
+  fileMetadata assigned: {'file-0': 'Class2.java', 'file-1': 'Class1.java', 'file-2': 'Class3.java'}.
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabFileHelper.js]: I'm not sure I understand this comment. Wouldn't it be 
`{'file-0': 'Class2.java', 'file-1': 'Class3.java', 'file-2': 'Class1.java'}`?

File: apps/src/javalab/JavalabFileHelper.js [hunk 5]
```diff
@@ -18,20 +18,80 @@ export function getDefaultFileContents(filename, viewMode) {
...
+  if (isValid) {
+    let count = 0;
+    let i = 0;
+    while (count < numFiles) {
+      let index = orderOfFiles.indexOf(i);
+      if (index > -1) {
+        orderedTabKeys.push(unorderedTabKeys[index]);
+        count++;
+      }
+      i++;
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabFileHelper.js]: index, count, and i are all often used as aliases for "array index." Give these more descriptive names indicating what they do in each loop.

File: apps/src/javalab/JavalabFileHelper.js [hunk 6]
```diff
@@ -42,6 +102,19 @@ export const fileMetadataForEditor = (sources, isEditingStartSources) => {
   };
 };
 
+export const updateAllSourceFileOrders = (
+  sources,
+  fileMetadata,
+  orderedTabKeys
+) => {
+  for (let i = 0; i < orderedTabKeys.length; i++) {
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabFileHelper.js]: It's ok to occasionally use `i` for an array index, but it's generally better to use a variable name that describes the purpose of `i`. 

Some context: we used to use a lot of abbreviations in code to decrease the number of keystrokes and the size of programs. But now that we have auto-complete and code minification, it's generally best practice to use full words to make the code more readable for the next developer. `i` is one of those abbreviations that has stuck around, but use it with care. Seeing it used in multiple places in this file starts to point to "code smell"

File: apps/src/javalab/JavalabFileHelper.js (hunk 3 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: I did consider changing the order in the json file, but I think that the order of object keys is not guaranteed. Also, when in levelbuilder startMode, validation and support files are added to sources for the levelbuilder and in the corresponding level for the student, these validation and support files are not included, but the other files would remain in the order set by the levelbuilder.

File: apps/src/javalab/JavalabFileHelper.js (hunk 4 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: In line 36, {'file-0': 'Class2.java', 'file-1': 'Class1.java', 'file-2': 'Class3.java'} is fileMetadata which is assigned as Object.keys iterates through the keys (file names) in sources and assigns the tabKey (metadata) names 'file-0', 'file-1', and 'file-2'. I chose an arbitrary order of keys within sources. 
Also while Object.keys iterates through the keys in sources, the orders of each file are added in array orderOfFiles while tabKey names are added to array unorderedTabKeys. 
From both of these arrays, we then get the orderedTabKeys. 
I was considering using the file names instead of the tabKey names for the ordering. However, the tabKeys are already being used in multiple Java Lab files.

File: apps/src/javalab/JavalabFileHelper.js (hunk 5 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: Great suggestion - will do!

File: apps/src/javalab/JavalabFileHelper.js (hunk 6 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: Thanks Jessie! This definitely makes sense.

File: apps/src/javalab/JavalabFileHelper.js (hunk 3 above)
Comment (jmkulwik) [reply apps/src/javalab/JavalabFileHelper.js]: Ahh ok. Taking this line of thinking just one step further. Would storing the files as a list rather than an object in the project json solve for this?

Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: Yes, if the files were stored in an array rather than an object, the logic to handle the ordering would have been more simple since order would've been guaranteed and we could rearrange order in the array when saving to sources. But it seems like this would require refactoring of other parts of Java Lab. 
Also, validation files are stored in a separate key from other files in the json object.

File: apps/src/javalab/JavalabEditor.jsx [hunk 7]
```diff
@@ -268,6 +271,15 @@ class JavalabEditor extends React.Component {
     };
   };
 
+  updateFileOrder() {
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabEditor.jsx:276]: nit: maybe `saveFileOrder`? `update` to me sounds like it's changing the order

File: apps/src/javalab/JavalabEditor.jsx [hunk 8]
```diff
@@ -389,11 +401,13 @@ class JavalabEditor extends React.Component {
     newTabs[index1] = newTabs[index2];
     newTabs[index2] = file1;
     setOrderedTabKeys(newTabs);
+    this.updateFileOrder();
     // closes the tab menu if it is open
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabEditor.jsx:399]: This is being handled by `updateFileOrder` now right?

File: apps/src/javalab/JavalabEditor.jsx [hunk 9]
```diff
@@ -389,11 +401,13 @@ class JavalabEditor extends React.Component {
...
     this.setState({
       showMenu: false,
       contextTarget: null
     });
+    this.props.closeEditorDialog();
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabEditor.jsx]: Just curious, why this change? Was there a situation where the the order could be changed while a dialog was open?

File: apps/src/javalab/JavalabEditor.jsx [hunk 10]
```diff
@@ -624,7 +636,8 @@ class JavalabEditor extends React.Component {
 
     setActiveTabKey(key);
     setOrderedTabKeys(newTabs);
-
+    this.updateFileOrder();
+    projectChanged();
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabEditor.jsx]: updateFileOrder calls projectChanged right?

File: apps/src/javalab/JavalabFileHelper.js [hunk 11]
```diff
@@ -18,20 +18,80 @@ export function getDefaultFileContents(filename, viewMode) {
...
+  Object.keys(sources).forEach(fileName => {
+    if (sources[fileName].isVisible || isEditingStartSources) {
+      let fileTabKey = getTabKey(fileIndex);
+      fileMetadata[fileTabKey] = fileName;
+      unorderedTabKeys.push(fileTabKey);
+      let tabOrder = sources[fileName].tabOrder;
+      // files that are stored may not currently have an tabOrder assigned so that
+      // order is undefined
+      if (Number.isInteger(tabOrder)) {
+        orderOfFiles.push(sources[fileName].tabOrder);
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabFileHelper.js]: nit: `orderOfFiles.push(tabOrder)`

File: apps/src/javalab/JavalabFileHelper.js [hunk 12]
```diff
@@ -18,20 +18,80 @@ export function getDefaultFileContents(filename, viewMode) {
...
-      // from the list, because we skip hidden files in the tabs.
-      const tabIndex = orderedTabKeys.length;
-      let tabKey = getTabKey(tabIndex);
-      fileMetadata[tabKey] = file;
-      orderedTabKeys.push(tabKey);
+  let unorderedTabKeys = [];
+  let orderOfFiles = [];
+  let fileIndex = 0;
+  let isValid = true;
+  Object.keys(sources).forEach(fileName => {
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabFileHelper.js:43]: Could we just construct the `orderedTabKeys` array as we iterate through the sources? `sources[fileName].tabOrder` is just an index right? Something like
```
unorderedTabKeys.push(fileTabKey)
If (Number.isInteger(tabOrder)) {
   orderedTabKeys[tabOrder] = fileTabKey;
} else { // invalid }
```
and then you can just filter out `undefined/null` values that would be present at any missing indices?

File: apps/src/javalab/JavalabEditor.jsx (hunk 7 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabEditor.jsx:276]: The updateFileOrder function calls on the action SOURCE_FILE_ORDER_UPDATED which does update the tabOrder of the files in sources according to the orderedTabKeys. 
I think I may also opt to be consistent with the other function names that are similar in function (such as updateVisibility, updateFileOrder, updateValidation) also call on actions that update and save the values within sources.

File: apps/src/javalab/JavalabEditor.jsx (hunk 9 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabEditor.jsx]: Ahhh...thanks for catching that. I will remove it - it is unnecessary - when I first implemented this feature, I saw the dialog box: 'Reload site? Changes you made may not be saved.' when I reordered the tabs and refreshed the browser. So I tried adding this command. 
But then when I went to production, I saw this dialog box appear appear and realized it was expected. But forgot to remove at that time! Thanks so much.

File: apps/src/javalab/JavalabEditor.jsx (hunk 8 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabEditor.jsx:399]: updateFileOrder calls on sourceFileOrderUpdated which updates the tabOrder of files in sources according to orderedTabKeys that was just modified by the swapping of index1 and index2.

File: apps/src/javalab/JavalabEditor.jsx (hunk 10 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabEditor.jsx]: yes - thanks for catching the redundant projectChanged call! Just removed.

File: apps/src/javalab/JavalabFileHelper.js (hunk 11 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: Much simpler - thanks!

File: apps/src/javalab/JavalabFileHelper.js (hunk 12 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js:43]: Thanks! This will definitely make it more simple to construct the orderedTabKeys taking into account missing tabOrders from the validation or support files. Will do!
I didn't realize you could assign an element to an out of bounds index in the array. Cool idea!

File: apps/src/javalab/JavalabEditor.jsx (hunk 7 above)
Comment (sanchitmalhotra126) [reply apps/src/javalab/JavalabEditor.jsx:276]: Ahh got it, makes sense!

File: apps/src/javalab/JavalabEditor.jsx (hunk 8 above)
Comment (sanchitmalhotra126) [reply apps/src/javalab/JavalabEditor.jsx:399]: Oh I meant closing the tab menu - looks like updateFileOrder does that internally right?

Comment (fisher-alice) [reply apps/src/javalab/JavalabEditor.jsx:399]: Oh yes! It does - thanks! Will remove.

File: apps/src/javalab/JavalabFileHelper.js (hunk 4 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: Update: tabKeyNames are assigned to orderedTabKeys as we iterate through keys in sources, removing need for unorderedTabKeys - see [Sanchit's suggestion](https://github.com/code-dot-org/code-dot-org/pull/48183#discussion_r978868244) below.

File: apps/src/javalab/JavalabEditor.jsx (hunk 7 above)
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabEditor.jsx:276]: nit: mind adding a comment here explaining what this does (updating the file order, saving the project, and closing the tab menu) since these are three distinct actions?

File: apps/src/javalab/JavalabFileHelper.js [hunk 13]
```diff
@@ -18,19 +18,72 @@ export function getDefaultFileContents(filename, viewMode) {
...
+    fileTabKey => fileTabKey !== undefined
+  );
+
+  // Check orderOfFiles for duplicates
+  isValid = isValid && new Set(orderOfFiles).size === orderOfFiles.length;
+
+  // If any of the visible file tabOrders were undefined or there were duplicate tabOrders,
+  // iterate through fileMetadata keys (fileTabKeys) and reassign tabOrders of all the files
+  // and add to orderedTabKeys the fileTabKeys
+  if (!isValid) {
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabFileHelper.js:80]: Rather than iterating again, could you build an `unorderedTabKeys` array while you're iterating through `sources` the first time, and then use that array in place of `orderedTabKeys` in the case that isValid is false?

File: apps/src/javalab/JavalabFileHelper.js [hunk 14]
```diff
@@ -18,19 +18,72 @@ export function getDefaultFileContents(filename, viewMode) {
...
+  For example, if a level had one validation file with tabOrder 0 and one starter
+  file with tabOrder 1, the orderedTabKeys would currently be [undefined, 1].
+  Filter out the undefined's next.
+  */
+  orderedTabKeys = orderedTabKeys.filter(
+    fileTabKey => fileTabKey !== undefined
+  );
+
+  // Check orderOfFiles for duplicates
+  isValid = isValid && new Set(orderOfFiles).size === orderOfFiles.length;
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabFileHelper.js]: Do you need `orderOfFiles` still? If it's just to check for duplicates, you can do that while assigning `orderedTabKeys[tabOrder] = ...` by checking for collisions

File: apps/src/javalab/JavalabFileHelper.js [hunk 15]
```diff
@@ -18,19 +18,72 @@ export function getDefaultFileContents(filename, viewMode) {
...
+  // If any of the visible file tabOrders were undefined or there were duplicate tabOrders,
+  // iterate through fileMetadata keys (fileTabKeys) and reassign tabOrders of all the files
+  // and add to orderedTabKeys the fileTabKeys
+  if (!isValid) {
+    fileIndex = 0;
+    orderedTabKeys = [];
+    Object.keys(fileMetadata).forEach(fileTabKey => {
+      orderedTabKeys.push(fileTabKey);
+      let fileName = fileMetadata[fileTabKey];
+      sources[fileName].tabOrder = fileIndex;
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabFileHelper.js]: Do you need to reassign `sources[filename].tabOrder` here? Doesn't look like the sources object is returned, so would this have any effect?

**Review by sanchitmalhotra126 — COMMENTED** [review]
A couple more comments on the file metadata logic, but looks good otherwise!

File: apps/src/javalab/JavalabEditor.jsx (hunk 7 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabEditor.jsx:276]: Definitely will do - thanks!

File: apps/src/javalab/JavalabFileHelper.js (hunk 13 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js:80]: Sure - sounds good and will do.

File: apps/src/javalab/JavalabFileHelper.js (hunk 14 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: Ahhh nice! Will refactor. Thanks!

File: apps/src/javalab/JavalabFileHelper.js (hunk 15 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: Oh right! The sources get updated in call to updateAllSourceFileOrders - thanks! Will remove this line.

Comment (fisher-alice) [reply apps/src/javalab/JavalabFileHelper.js]: Actually, the line is already replaced by the change from above: https://github.com/code-dot-org/code-dot-org/pull/48183#discussion_r980447836

File: apps/src/javalab/JavalabFileHelper.js [hunk 16]
```diff
@@ -18,19 +18,68 @@ export function getDefaultFileContents(filename, viewMode) {
...
+  let isValid = true;
+  Object.keys(sources).forEach(fileName => {
+    if (sources[fileName].isVisible || isEditingStartSources) {
+      let fileTabKey = getTabKey(fileIndex);
+      fileMetadata[fileTabKey] = fileName;
+      let tabOrder = sources[fileName].tabOrder;
+      // Files that are stored may not currently have an tabOrder assigned so that
+      // order is undefined
+      // Also check if two files have the same tabOrder. If so, then orderedTabKeys is invalid
+      if (
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabFileHelper.js:51]: nit: you could also check if `isValid` is already false here to short-circuit this condition

File: apps/src/javalab/JavalabEditor.jsx (hunk 8 above)
Comment (fisher-alice) [reply apps/src/javalab/JavalabEditor.jsx:399]: Update: Removed the closing of the tab menu from updateFileOrder to narrow scope of what updateFileOrder does.

**Review by sanchitmalhotra126 — APPROVED** [review]
One small nit, but LGTM otherwise! Nice work on this!
